### PR TITLE
Clarify how *ref-safe-to-escape*/*safe-to-escape* are calculated for inline array operations.

### PR DESCRIPTION
**Pull request by AlekseyTs**

### CODE REVIEW

File: proposals/inline-arrays.md [hunk 1]
```diff
@@ -113,17 +113,20 @@ For an inline array element access, the *primary_no_array_creation_expression* o
...
-method on *primary_no_array_creation_expression*. 
+method on *primary_no_array_creation_expression*. For the purpose of ref-safety analysis the *ref-safe-to-escape*/*safe-to-escape*
+of the access are equivalent to the same for an invocation of a method with the signature ```ref readonly T GetItem(in InlineArrayType array)```. 
 
 If *primary_no_array_creation_expression* is a value, the result of evaluating an inline array element access is a value
 equivalent to invoking [`public ref readonly T this[int index] { get; }`](https://learn.microsoft.com/en-us/dotnet/api/system.readonlyspan-1.item?view=net-8.0) with
 that integer value on an instance of ```System.ReadOnlySpan<T>``` returned by ```System.ReadOnlySpan<T> InlineArrayAsReadOnlySpan```
-method on *primary_no_array_creation_expression*. 
+method on *primary_no_array_creation_expression*. For the purpose of ref-safety analysis the *ref-safe-to-escape*/*safe-to-escape*
+of the access are equivalent to the same for an invocation of a method with the signature ```T GetItem(in InlineArrayType array)```.
```
Comment (AlekseyTs) [inline proposals/inline-arrays.md:129]:
```suggestion
of the access are equivalent to the same for an invocation of a method with the signature ```T GetItem(InlineArrayType array)```. 
```

File: proposals/inline-arrays.md [hunk 2]
```diff
@@ -113,17 +113,20 @@ For an inline array element access, the *primary_no_array_creation_expression* o
 
 If *primary_no_array_creation_expression* is a writable variable, the result of evaluating an inline array element access is a writable variable
 equivalent to invoking [`public ref T this[int index] { get; }`](https://learn.microsoft.com/en-us/dotnet/api/system.span-1.item?view=net-8.0) with
-that integer value on an instance of ```System.Span<T>``` returned by ```System.Span<T> InlineArrayAsSpan``` method on *primary_no_array_creation_expression*. 
+that integer value on an instance of ```System.Span<T>``` returned by ```System.Span<T> InlineArrayAsSpan``` method on *primary_no_array_creation_expression*. For the purpose of ref-safety analysis the *ref-safe-to-escape*/*safe-to-escape*
+of the access are equivalent to the same for an invocation of a method with the signature ```ref T GetItem(ref InlineArrayType array)```.
```
Comment (jaredpar) [inline proposals/inline-arrays.md:117]:
```suggestion
of the access are equivalent to the same for an invocation of a method with the signature ```static ref T GetItem(ref InlineArrayType array)```.
```

Consider using `static` here and the other methods in this PR to make it clear the receiver is being passed explicitly vs. implicitly via `this`. That distinction is important for ref safety rules.

**Review by jaredpar — APPROVED** [review]

File: proposals/inline-arrays.md [hunk 3]
```diff
@@ -113,17 +113,20 @@ For an inline array element access, the *primary_no_array_creation_expression* o
...
 
 If *primary_no_array_creation_expression* is a readonly variable, the result of evaluating an inline array element access is a readonly variable
 equivalent to invoking [`public ref readonly T this[int index] { get; }`](https://learn.microsoft.com/en-us/dotnet/api/system.readonlyspan-1.item?view=net-8.0) with
 that integer value on an instance of ```System.ReadOnlySpan<T>``` returned by ```System.ReadOnlySpan<T> InlineArrayAsReadOnlySpan```
-method on *primary_no_array_creation_expression*. 
+method on *primary_no_array_creation_expression*. For the purpose of ref-safety analysis the *ref-safe-to-escape*/*safe-to-escape*
+of the access are equivalent to the same for an invocation of a method with the signature ```ref readonly T GetItem(in InlineArrayType array)```.
```
Comment (AlekseyTs) [inline proposals/inline-arrays.md:123]:
```suggestion
of the access are equivalent to the same for an invocation of a method with the signature ```static ref readonly T GetItem(in InlineArrayType array)```. 
```

File: proposals/inline-arrays.md [hunk 4]
```diff
@@ -113,17 +113,20 @@ For an inline array element access, the *primary_no_array_creation_expression* o
...
-method on *primary_no_array_creation_expression*. 
+method on *primary_no_array_creation_expression*. For the purpose of ref-safety analysis the *ref-safe-to-escape*/*safe-to-escape*
+of the access are equivalent to the same for an invocation of a method with the signature ```ref readonly T GetItem(in InlineArrayType array)```. 
 
 If *primary_no_array_creation_expression* is a value, the result of evaluating an inline array element access is a value
 equivalent to invoking [`public ref readonly T this[int index] { get; }`](https://learn.microsoft.com/en-us/dotnet/api/system.readonlyspan-1.item?view=net-8.0) with
 that integer value on an instance of ```System.ReadOnlySpan<T>``` returned by ```System.ReadOnlySpan<T> InlineArrayAsReadOnlySpan```
-method on *primary_no_array_creation_expression*. 
+method on *primary_no_array_creation_expression*. For the purpose of ref-safety analysis the *ref-safe-to-escape*/*safe-to-escape*
+of the access are equivalent to the same for an invocation of a method with the signature ```T GetItem(InlineArrayType array)```.
```
Comment (AlekseyTs) [inline proposals/inline-arrays.md:129]:
```suggestion
of the access are equivalent to the same for an invocation of a method with the signature ```static T GetItem(InlineArrayType array)```. 
```

File: proposals/inline-arrays.md [hunk 5]
```diff
@@ -171,12 +174,16 @@ the *primary_no_array_creation_expression*. Then the element access is interpret
 
 If *primary_no_array_creation_expression* is a writable variable, the result of evaluating an inline array element access is a value
 equivalent to invoking [`public Span<T> Slice (int start, int length)`](https://learn.microsoft.com/en-us/dotnet/api/system.span-1.slice?view=net-8.0)
-on an instance of ```System.Span<T>``` returned by ```System.Span<T> InlineArrayAsSpan``` method on *primary_no_array_creation_expression*. 
+on an instance of ```System.Span<T>``` returned by ```System.Span<T> InlineArrayAsSpan``` method on *primary_no_array_creation_expression*.
+For the purpose of ref-safety analysis the *ref-safe-to-escape*/*safe-to-escape* of the access are equivalent to the same
+for an invocation of a method with the signature ```System.Span<T> GetSlice(ref InlineArrayType array)```.
```
Comment (AlekseyTs) [inline proposals/inline-arrays.md:179]:
```suggestion
for an invocation of a method with the signature ```static System.Span<T> GetSlice(ref InlineArrayType array)```.
```

File: proposals/inline-arrays.md [hunk 6]
```diff
@@ -171,12 +174,16 @@ the *primary_no_array_creation_expression*. Then the element access is interpret
...
 
 If *primary_no_array_creation_expression* is a readonly variable, the result of evaluating an inline array element access is a value
 equivalent to invoking [`public ReadOnlySpan<T> Slice (int start, int length)`](https://learn.microsoft.com/en-us/dotnet/api/system.readonlyspan-1.slice?view=net-8.0)
 on an instance of ```System.ReadOnlySpan<T>``` returned by ```System.ReadOnlySpan<T> InlineArrayAsReadOnlySpan```
-method on *primary_no_array_creation_expression*. 
+method on *primary_no_array_creation_expression*.
+For the purpose of ref-safety analysis the *ref-safe-to-escape*/*safe-to-escape* of the access are equivalent to the same
+for an invocation of a method with the signature ```System.ReadOnlySpan<T> GetSlice(in InlineArrayType array)```.
```
Comment (AlekseyTs) [inline proposals/inline-arrays.md:186]:
```suggestion
for an invocation of a method with the signature ```static System.ReadOnlySpan<T> GetSlice(in InlineArrayType array)```.
```

File: proposals/inline-arrays.md [hunk 7]
```diff
@@ -243,6 +250,10 @@ void M3()
 }
 ```
 
+For the purpose of ref-safety analysis the *safe-to-escape* of the conversion is equivalent to *safe-to-escape*
+for an invocation of a method with the signature ```System.Span<T> Convert(ref InlineArrayType array)```, or
+```System.ReadOnlySpan<T> Convert(in InlineArrayType array)```.
```
Comment (AlekseyTs) [inline proposals/inline-arrays.md:255]:
```suggestion
```static System.ReadOnlySpan<T> Convert(in InlineArrayType array)```.
```

**Review by jaredpar — APPROVED** [review]

File: proposals/inline-arrays.md [hunk 8]
```diff
@@ -113,17 +113,20 @@ For an inline array element access, the *primary_no_array_creation_expression* o
 
 If *primary_no_array_creation_expression* is a writable variable, the result of evaluating an inline array element access is a writable variable
 equivalent to invoking [`public ref T this[int index] { get; }`](https://learn.microsoft.com/en-us/dotnet/api/system.span-1.item?view=net-8.0) with
-that integer value on an instance of ```System.Span<T>``` returned by ```System.Span<T> InlineArrayAsSpan``` method on *primary_no_array_creation_expression*. 
+that integer value on an instance of ```System.Span<T>``` returned by ```System.Span<T> InlineArrayAsSpan``` method on *primary_no_array_creation_expression*. For the purpose of ref-safety analysis the *ref-safe-to-escape*/*safe-to-escape*
+of the access are equivalent to the same for an invocation of a method with the signature ```static ref T GetItem(ref InlineArrayType array)```.
```
Comment (RikkiGibson) [inline proposals/inline-arrays.md:117]: Was the original problem that the `this` parameter is implicitly `scoped ref`, but the element access is supposed to return a ref to the element within `this`? And I guess a similar issue in the other changed sections?

**Review by RikkiGibson — APPROVED** [review]

File: proposals/inline-arrays.md (hunk 8 above)
Comment (jaredpar) [reply proposals/inline-arrays.md:117]: Correct.

**Pull request merged**